modules/dhcp_common/subnet_service.rb

```diff
@@ -15,15 +18,18 @@ def initialize(subnets, leases_by_ip, leases_by_mac, reservations_by_ip, reserva
     end
 
     def self.initialized_instance
-      new(::Proxy::MemoryStore.new, ::Proxy::MemoryStore.new, ::Proxy::MemoryStore.new,
+      new(::Proxy::MemoryStore.new, ::Proxy::MemoryStore.new,
           ::Proxy::MemoryStore.new, ::Proxy::MemoryStore.new, ::Proxy::MemoryStore.new)
     end
 
     def add_subnet(subnet)
       m.synchronize do
-        raise Proxy::DHCP::Error, "Unable to add subnet #{subnet}" if find_subnet(subnet.network)
+        key = ipv4_to_i(subnet.network) & ipv4_to_i(subnet.netmask)
```

Comment:
`ipv4_to_i(subnet.network)` can be replaced by `subnet.ipaddr.to_i` (if you add an attr_reader), which seems to have a small performance benefit here. I think IPAddr already calculates this value, and ipv4_to_i has a slight cost.

Base:
<pre>
Calculating -------------------------------------
      add_subnet (1)     31.170k (±17.0%) i/s -    288.249k in   9.820534s
      add_subnet (5)      7.418k (±12.6%) i/s -     71.478k in   9.938092s
     add_subnet (50)    767.327  (±10.4%) i/s -      7.548k in   9.992509s
    add_subnet (500)     74.870  (± 9.3%) i/s -    742.000  in  10.011340s
   add_subnet (1000)     37.684  (± 8.0%) i/s -    375.000  in  10.009325s
  add_subnet (15000)      2.304  (± 0.0%) i/s -     23.000  in  10.058093s
Memory stats
Total objects allocated: 77750679
</pre>

Changed:
<pre>
Calculating -------------------------------------
      add_subnet (1)     33.397k (±17.1%) i/s -    308.262k in   9.820294s
      add_subnet (5)      8.128k (±12.0%) i/s -     78.568k in   9.937083s
     add_subnet (50)    845.850  (± 8.6%) i/s -      8.355k in   9.993047s
    add_subnet (500)     82.899  (± 6.0%) i/s -    825.000  in   9.999682s
   add_subnet (1000)     40.876  (± 7.3%) i/s -    407.000  in  10.017958s
  add_subnet (15000)      2.575  (± 0.0%) i/s -     26.000  in  10.158017s
Memory stats
Total objects allocated: 66702210
Total heap pages allocated: 760
</pre>